### PR TITLE
Show modified bases using MM and MP/ML tags in BAM/CRAM

### PR DESCRIPTION
**Pull request by cmdcolin**
Stub for coloring reads that have MM (positions of modified bases) and MP/ML (modification likelihood/probability) tags

This can be used to show methylation and other types of modifications, but methylation is probably the most common application

https://github.com/GMOD/jbrowse/issues/1588

Needs to still incorporate CIGAR for positioning, but this is a fun start

Requested here https://github.com/GMOD/jbrowse/issues/1588

### CODE REVIEW

**Comment by cmdcolin** [comment]
Note that to do this properly, it is probably useful for SNPCoverage to also show these modification in the pileup too

**Comment by cmdcolin** [comment]
Other todos: 
- incorporate into snpcoverage renderer/adapter
- dynamically set color palette for different modifications (similar to our color by tag functionality)
- allow only drawing a single modification type at a time perhaps
- possibly draw multiple modifications at a single base pair

**Comment by cmdcolin** [comment]
possibly also making the wgbs type display could be worth doing in paralell, as some modern tools for calling modifications output pseudo-wgbs mode bam files https://software.broadinstitute.org/software/igv/interpreting_bisulfite_mode

**Comment by cmdcolin** [comment]
Some initial results here with both SNPCoverage and Pileup track displaying modifications, incorporating CIGAR indels to make sure modifications are in the right position on the reference

This requires some rewriting of SNPCoverage to allow it to support displaying modifications. It may end up that we refactor mismatches in a similar way to what is done here with modification positions

localhost_3000__config=test_data%2Fconfig_demo json session=local-lB9UrwLd9

**Comment by cmdcolin** [comment]
Only one type of modification visible in this data set that i've seen so far, so just red...

**Comment by cmdcolin** [comment]
Added a specialized "color by methylation" setting to try to give a sense of whether a particular base is modified or not. In general, with DNA methylation, we often look at "CpG" elements which are the places where it is the literal letters "CG" in the genome e.g. C followed by G, which are so called CpG's. The C part of the string is often methylated, so it is useful to color the CpG one color if they are methylated, and another color of they are not. If you don't color the ones that aren't methylated, you lose the "perspective" on the methylation proportions in a region.

Here is a screenshot with this setting on, using the sample data

localhost_3000__config=test_data%2Fconfig_demo json session=local-1zJZnWqPE (1)

**Comment by cmdcolin** [comment]
This could tentatively be marked as ready. It is pretty basic still, but may satisfy basic needs and iterated on

Has a two modes, methylation and modification

- methylation setting (looking at instances of CpG in the reference genome [a CpG is a C followed by a G] and checking whether the MM tag marks a modification on the C, which is a common form of methylation). this could be elaborated to show multiple types of methylation, not just CpG (e.g. CHH "In plants and other organisms, DNA methylation is found in three different sequence contexts: CG (or CpG), CHG or CHH (where H correspond to A, T or C). In mammals however, DNA methylation is almost exclusively found in CpG dinucleotides" https://en.wikipedia.org/wiki/DNA_methylation)
- a modifications setting (marks raw modifications seen in the MM tag, shaded by ML tag), this could be elaborated on to add configurable colors for different modification types, etc.



Interesting notes:

* In both SNPCoverageAdapter and PileupRenderer (yes renderer) it manually obtains the sequence adapter using readConfObject(dataAdapter, ['sequenceAdapter']). This is because they need to read the reference sequence underlying the read to mark CpG's accurately.
* It could be argued that to mark CpG it should look at the read itself instead of the underlying reference sequence for a C following a G but the reads are so noisy almost 90% mismatches after calmd used in the nanopore example that it is not trustworthy
* we could add some more work to adopt more visualizations from things like bedMethyl, or jb1 methylation style https://twitter.com/methylcytosine/status/1389575911343800320, or WGBS (whole genome bisulfite sequencing bam)

**Comment by cmdcolin** [comment]
note that the SNPCoverageAdapter now has "special modes", e.g. it gets a colorBy flag via the renderProps getter as the second argument to getFeatures, and so it calculates per-base coverage of the modifications, methylation, or the normal mismatches

**Review by rbuels — COMMENTED** [review]
looks good, just a couple little questions

File: plugins/alignments/package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
     "abortable-promise-cache": "^1.1.3",
     "color": "^3.1.2",
     "copy-to-clipboard": "^3.3.1",
+    "d3-interpolate": "^2.0.1",
```
[the remark this answers is not in the document]
Comment (cmdcolin) [reply plugins/alignments/package.json:44]: it was used to interpolate color from white -> modification color, but alpha fade out may be better anyways. removed

File: plugins/alignments/src/SNPCoverageRenderer/SNPCoverageRenderer.test.js [hunk 2]
```diff
@@ -1,68 +0,0 @@
...
-import SNPCoverageRenderer, { configSchema, ReactComponent } from '.'
-
-test('several features', async () => {
-  const pluginManager = {}
-  const renderer = new SNPCoverageRenderer({
-    name: 'SNPCoverageRenderer',
-    ReactComponent,
-    configSchema,
-    pluginManager,
-  })
```
[the remark this answers is not in the document]
Comment (cmdcolin) [reply plugins/alignments/src/SNPCoverageRenderer/SNPCoverageRenderer.test.js:12]: the SNPCoverageRenderer is pretty well covered by integration tests and would need updating about the format of features it expects to get, the format for the feature.get('snpinfo') changed in this PR to be more purpose built to handle things other than snps

**Comment by garrettjstevens** [comment]
image

image

**Pull request merged**